[Virtual DOM 만들기] 피터(안현철) 미션 제출합니다.

iborymagic opened the pull request:
안녕하세용 콜린~!
이렇게 하는 게 맞는건가 싶긴 하지만, 일단 동작은 하더라구요.
그래서 일단 제출해놓고 좀 더 고쳐 볼 예정입니다.

현재는 업데이트가 될 때마다 root를 전부 비우고 다시 채워넣고 있어요.
즉, diff 알고리즘 같은 건 아직 구현하지 못했습니다. ㅎㅎ
모든 요구사항을 만족시키지 못한 대신, 최대한 알아보기 쉬운 코드를 작성하려고 노력해봤어요.
가독성이 떨어지거나 하는 코드가 있으면, 가차없이 지적해주시면 감사할 것 같습니다.

일단 PR을 제출했으니, 제 나름대로 diff 알고리즘을 구현하는 것이 이후의 가장 큰 목표이고
가능하다면 parser까지 만들어보고 싶긴 하네요.
잘 부탁드립니다.

[데모 사이트](https://iborymagic.github.io/javascript-own-ui-library/)

- [x] VirtualDOM 생성 팩토리 함수 구현 (aka createElement)
- [x] 렌더러 구현
- [x] 직접 구현한 렌더러와 VirtualDOM으로 Counter App 렌더링
- [x] DOM Node 비교
- [x] 업데이트 DOM Node
- [x] 렌더링한 Counter App 동작시키기


> 코드 설명
### 팩토리 함수 
createElement() - Peact.js
- 아직은 parser를 구현하지 못해서, App.js에 가보시면 return부가 createElement로 되어있는 모습을 보실 수 있습니다. ㅎ
- createElement 함수는 말 그대로 팩토리 함수입니다. 
- tag의 type과 props, 그리고 children을 인자로 받으면, 그것을 이용해 제 나름대로 'Virtual DOM'이라고 정의한,
렌더링 될 element의 성질들을 담고 있는 객체를 생성합니다.
- 코드를 보시면 children이 함수 타입일 때는 실행해주고, 그게 아니면 그대로 반환해주는 모습을 보실 수 있을 거에요.
그렇게 하면 text node를 만날 때까지 함수를 재귀적으로 실행하면서 쭉쭉 내려가게 됩니다.
(text node는 type으로 'text'를 받습니다)

### 렌더러 함수
1. render() - PeactDOM.js
- 처음 렌더링을 하는 함수입니다.
- 아직 diff 알고리즘을 만들진 않았지만, 후에 목표로 하고 있기 때문에 vDOM을 'prevSnapshot'이라는 객체에 저장하고 있습니다.
- 그리고 makeDOM함수를 실행하여, createElement를 통해 생성된 vDOM을 순회하며 root에 element를 append 해줍니다.
2. makeDOM() - PeactDOM.js
- 역시나 재귀 함수입니다.
- 객체를 맨 위에서부터 차례차례 한 꺼풀씩 구조 분해 할당으로 벗겨내면서 순회합니다.
- 재귀의 종료 조건은 'text node'를 만나면 종료됩니다.
- createElement로 element를 생성하고, attributes를 붙여준 후에 인자로 받은 parent에 appendChild를 합니다.
- props(객체)를 순회하며, function이면 eventlistener로 인식하여 addEventListener를 실행하고, 
아니면 setAttribute를 실행합니다.
- appendChild를 한 후에, children을 돌면서 다시 makeDOM 함수를 호출해줍니다. (재귀)
3. reRender() - PeactDOM.js
- 리렌더링 함수인데, 아직까지는 diff 알고리즘을 구현하지 못해서 그냥 root를 비워주고 싹 다 리렌더링하고 있습니다.
- setState 함수가 호출될 때마다 reRender 함수도 함께 호출됩니다.

### 상태 관리
- 상태 관리는 State라는 클래스에 stateIndex와 stateTable을 만들어서 하고 있습니다.
- stateIndex를 키로 해서 { state, setState }를 저장하는 식입니다.
- 특정 컴포넌트 내에서는 useState가 항상 동일한 순서로 호출된다고 생각하고 만들었습니다.
- 가장 처음에 위치한 useState가 호출될 때는 stateIndex가 0입니다.
- stateIndex는 useState가 호출될 때마다 1씩 증가합니다.
- 그래서 매번 리렌더링될 때마다 stateIndex를 통해 자기 자신의 state를 찾아가는 방식입니다.
- setState를 호출할 때마다 reRender를 호출, 즉 리렌더링이 됩니다. 그 때마다 stateIndex도 0으로 초기화시켜줍니다.

이상입니다. 
혹시 이해가 어려운 부분이 있으면 언제든지 연락주세용
감사합니다!

iborymagic committed:
feat: createElement 구현

iborymagic committed:
feat: render 구현

iborymagic committed:
feat: useState 구현

iborymagic committed:
feat: render 함수 구조 변경
- 시작 함수와 재귀 함수를 분리
- documentFragment를 이용하도록 수정

iborymagic committed:
feat: state와 useState 구현

iborymagic committed:
refactor: Counter 컴포넌트를 App 컴포넌트에 삽입

iborymagic committed:
fix: render 함수에서 Component 대신 Virtual DOM 객체를 인자로 받도록 수정

iborymagic committed:
fix: diff 알고리즘 함수 제거

2SOOY commented on scenario.txt:
```diff
@@ -0,0 +1,79 @@
+// HTML string => parser => createElement => virtual DOM object => render => real DOM
+
```
파일 이름이 굉장히 적절하군요 :) 
HTML 문자열에서 VDOM을 거쳐 real DOM까지 변환되는 과정을 써주셔서 생각의 흐름을 볼 수 있어서 좋네용
구현할 때까지 응원하겠습니다 ✊✊✊

2SOOY commented on lib/Peact.js:
```diff
@@ -0,0 +1,11 @@
+  return {
+    type,
+    props,
+    children: children.map((child) =>
+      typeof child === 'function' ? child() : child
+    ),
```
`typeof child === 'function'`일 때의 사용 케이스도 보고 싶네용 :)
피터가 작성해주신 코드를 바탕으로 만약 prop을 넘겨줘야하는 상황이라면 다음과 같은 방식으로 사용하면 될까용?

```js
// 컴포넌트 선언부
const Child = ({ p1, p2 }) => {
 return createElement('tag', {p1, p2})
}

// 컴포넌트 사용부
const Parent = () => {
   return createElement('tag', {}, 
        () => Child({ p1, p2 }), 
        () => Child({ p1, p2 }))
}
```

2SOOY replied:
index.js에서 App 컴포넌트 자체를 호출한 것 처럼 나머지 컴포넌트 사용부를 다음과 같이 통일하는 방식은 어떤가용?
```js
// 컴포넌트 사용부
const Parent = () => {
   return createElement('tag', {}, Child({ p1, p2 }), Child({ p1, p2 }))
}
```

iborymagic replied:
제가 콜린의 질문을 정확히 이해한 게 맞다면, 
요 부분은 JSX 파서를 만든다는 가정 하에 App 컴포넌트의 return 부가
```jsx
<div class="container">
  <span class="count">0</span>
  <div class="btn-group">
    <button>
      <strong>-</strong>
    </button>
    <button>
      <strong>RESET</strong>
    </button>
    <button>
      <strong>+</strong>
    </button>
  </div>
</div>
```
요렇게 생겼다고 생각하고 만든 것이랍니당
JSX 파서로 태그를 변환한 결과물이 createElement라는 생각이에요.
그래서 HTML 태그들은 따로 컴포넌트로 분리하지않고, 그냥 createElement를 쌓아놓는 식으로 해뒀어요.
근데, 각 보니 JSX 파서는 좀 힘들 것 같기도..ㅎㅎ;;

2SOOY replied:
> HTML 태그들은 따로 컴포넌트로 분리하지않고, 그냥 createElement를 쌓아놓는 식으로 해뒀어요.

이해했습니다 :) 나중에 여유로울 때 하시고 알려주세용 🚀

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,47 @@
+const prevSnapshot = { vDOM: null };
+
```
`snapshot` 마음에 드는 변수명이네용 👍

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,47 @@
+  if (type === 'text') {
+    parent.innerHTML = children[0];
+    return;
+  }
```
`textNode`일 때는 `innerHTML`대신 `parent.textContent`는 어떨까용?

iborymagic replied:
오! 그렇네요
무심코 넘어갔던 부분!

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,47 @@
+  // parse => createElement(...)
+  prevSnapshot.vDOM = vDOM;
+
+  const fragment = document.createDocumentFragment();
+  makeDOM(vDOM, fragment);
+  root.appendChild(fragment);
```
`fragment`를 사용하신 이유가 궁금합니다 ! 그리고 질문하나 있습니당 :)

실제 브라우저에 렌더링 되지 않은 `DOM 요소`에 appendChild와 같은 DOM API를 사용하는 상황에서
1. 해당 `DOM 요소`가 `fragment`일 때,
2. 해당 `DOM 요소`가 일반 `HTMLElement`일 때, 

아직은 렌더링 되지 않은 `DOM 요소`를 조작하는데도 성능 차이가 존재할까용 ? 
(제가 frgment를 사용해 본 적이 없어서 궁금하네용)

iborymagic replied:
둘의 차이점은 최상위 요소가 document에 반영되느냐 아니냐의 차이밖에 없는걸로 알고 있어용
그래서 별도의 컨테이너 없이 여러 개의 node들을 부모 요소에 한 번에 붙이고 싶은 경우에 
documentFragment가 유용한 걸로 알고 있습니다. 

그냥 HTMLElement이면 여러 node를 붙이기 위해 DOM에 여러 번 접근하거나,
span이나 div같은 부모 컨테이너를 하나 만들어야 그 children으로 여러 개의 node들을 붙일 수 있으니까요.
(말씀하신 실제 브라우저에 렌더링되지 않은 DOM 요소 중의 최상위 요소가 이 컨테이너겠네요)

그런 경우를 제외하곤, 성능 차이는 딱히 없는 걸로 알고있는데...
사실 아무리 찾아봐도 '둘이 성능 차이가 이만큼 있다!' 라고 정확히 명시한 사람을 못봐서 그렇다고 생각하는거지, 
확실하진 않슴다.

제 코드에서는 재귀적으로 appendChild를 해주되, 
컨테이너를 만들지 않으면서 DOM에 여러 번 접근하지 않기 위해 fragment를 사용했답니당

2SOOY replied:
> 컨테이너를 만들지 않으면서 DOM에 여러 번 접근하지 않기 위해 fragment를 사용했답니당

그렇군용 !! 상세한 이유 답변해 주셔서 감사합니다 :)

2SOOY commented on lib/useState.js:
```diff
@@ -0,0 +1,39 @@
+import App from '../App.js';
+import { reRender } from './PeactDOM.js';
+
+class State {
+  #stateIndex;
+  #stateTable;
```
`step2`도 미리 구현해주셨군요 👍👍👍

2SOOY commented on lib/useState.js:
```diff
@@ -0,0 +1,39 @@
+    // state를 setState에서 변경 가능한 상태로 가둬놓기 위해 객체로 선언
+    const stateObj = {
+      state: this.#stateTable?.get(this.#stateIndex)
+        ? this.#stateTable?.get(this.#stateIndex).state
```
`this.#stateTable?.get(this.#stateIndex)` 옵셔널 체이닝으로 `#stateTable`에 접근하는 이유가 궁금합니다 :)
`State` 클래스는 반드시 instance화 시켜서 사용할텐데, `constructor()` 내부에서 `#stateTable`이랑 `#stateIndex`를 초기화 해주지 않나용?

iborymagic replied:
앗 ㅎㅎ 
부끄럽지만, 호다닥 작성하다가
습관처럼 아무 생각 없이 사용한 옵셔널 체이닝이었네요...
삭제하도록 하겠습니다..

2SOOY commented on lib/useState.js:
```diff
@@ -0,0 +1,39 @@
+      state: this.#stateTable?.get(this.#stateIndex)
+        ? this.#stateTable?.get(this.#stateIndex).state
+        : value,
+      setState: (newValue) => {
+        stateObj.state = newValue;
+        this.#resetIndex();
+        reRender(App());
+      },
```
리렌더링 과정에서 모든 함수 컴포넌트가 호출되고 컴포넌트 내부의 useState역시 호출되고 또한 각 state간의 호출 순서를 보장하기 위해서 index를 초기화 하는 과정이 필요하군용 👍👍👍 `useState` 함수를 멋지게 구현하셨네용 :)

개인적으로 이를 어떻게 구현해야할지 아이디어가 떠오르지 않았는뎅 참고하겠습니다 😃

2SOOY reviewed (changes requested):
안녕하세요! 피터 :) 🕸️  

우선 바쁜 와중에도 미션 구현하시느라 고생하셨습니다 👏👏👏
 
README에 피터가 구현한 함수 각각에 대한 설명 덕분에 코드 이해하기가 훨씬 수월해서 감사했습니다 👍 가상돔 뿐만 아니라 `STEP2`에 해당하는 상태 관리도 구현하셨다니 대단해용 👍👍👍 피터 코드 보면서 제 나름대로 학습의 시간을 가질 수 있었어요 !!

간단하게 코멘트 남겼으니 확인 부탁드려용 :) 이후 `diff` 관련하여 업데이트 진행되면 지속적으로 이야기 나눠보면 좋을 것 같아요 !!

iborymagic committed:
feat: diff 알고리즘(findDifference 함수) 구현

iborymagic committed:
docs: readme.md 필수 요구사항 체크

iborymagic committed:
fix: text node일 때 innerHTML 대신 textContent를 사용

iborymagic committed:
refactor: 불필요한 optional chaining 삭제

iborymagic committed:
refactor: 불필요한 변수 changed 삭제

iborymagic committed:
fix: findDifference 함수에서 props 비교 시 optional chaining으로 방어

iborymagic commented:
findDifference 함수를 구현했습니다.
단순히 이전 vDOM, 새로운 vDOM, 그리고 realDOM을 최상위부터 순서대로 탐색하면서 type, props를 비교하는 함수입니다.
만약 차이점이 발견되면 새로운 vDOM을 기반으로 element를 만들어서 해당 부분의 realDOM과 교체합니다.

조금 독특한 부분이라면 props가 function인 경우, 즉 eventListener가 달린 경우인데,
두 함수를 string으로 변환해서 비교하고, 만약 차이점이 발견된다면 역시나 element를 만들고 realDOM을 교체합니다.
차이점이 발견되지 않는 경우 realDOM에서 removeEventListener를 하고나서 addEventListener를 해주고 있습니다.

remove를 먼저 해주고나서 add를 하는 이유는, 
findDifferrence 함수는 필요한 부분만 새로 갈아끼워주는 함수인데, 그러면 새로 갈아끼워지지 않은 
나머지 부분의 eventListener들은 이전 것들, 즉 바뀌기 전의 상태값을 가진 함수들이기 때문이에요.
그 부분들까지 전부 새로운 상태값을 가진 함수들로 갈아끼워주기 위해서
변경이 되지 않더라도 function type이면 전부 newVDOM의 것들로 갈아끼워주고 있습니다.

eventListener가 변경되는 경우에 remove 후 add를 해주지 않는 이유는, 
eventListener가 새로 생성될 수도 있고 기존의 eventListener가 사라질 수도 있어서
그냥 아예 해당 element와 children들을 새로 갈아버리도록 구현했습니다!

주석으로 최대한 설명하려고 노력했는데,
혹시 이해가 안되는 부분이 있으면 언제든 물어봐주십쇼!

ps. 파서는 아직..ㅎㅎ;

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,114 @@
+    }
+  });
+}
+
+function replaceElement(content, target) {
+  const fragment = document.createDocumentFragment();
+  makeDOM(content, fragment);
+  target.replaceWith(fragment);
+}
+
```
질문 ! 호이스팅을 활용하려고 `function`을 사용하신건가용?

iborymagic replied:
맞아요! 
코드 리뷰를 위해서 최대한 흐름이 자연스럽게 함수를 배치하려고 했는데, 
함수 위치를 이리저리 옮길 때 귀찮지 않기 위해서 그냥 function을 냅다 갈겼습니다.

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,114 @@
+
+/* re-rendering 함수 */
+function reRender(newVDOM) {
+  const rootElement = document.getElementById('root');
+  const appComponent = rootElement.children[0];
```
```suggestion
  const rootElement = document.getElementById('root');
  const appComponent = rootElement.firstChild;
```
첫 번째 자식 요소를 얻고 싶다면 조금 더 선언적으로 얻는 방법도 있을 것 같아용 :)

iborymagic replied:
오 좋습니다!!

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,114 @@
+  prevSnapshot.vDOM = newVDOM;
+}
+
+/* 이전 vDOM과 새로운 vDOM의 차이를 찾아내 replace 하는 함수 */
+function findDifference(prevVDOM, newVDOM, currentRealDOM) {
+  const { type: prevType, props: prevProps, children: prevChildren } = prevVDOM;
```
소소한 제안 :)
해당 함수에서 비교 뿐만 아니라 DOM을 교체하는 작업까지 담당하고 있으니깐 `find`보다는 `apply-` or `update-` 정도의 네이밍은 어떨까용?

iborymagic replied:
좀 더 역할에 맞는 명확한 네이밍이네요!!
좋은 제안 감사합니다 :)

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,114 @@
+  const propsKeys =
+    newPropsKeys.length > prevPropsKeys.length ? newPropsKeys : prevPropsKeys;
+
+  // 종료 조건
+  if (!newChildren.length) {
+    return;
+  }
+
```
질문이 있습니당 :)

`newVDOM`이 `children`이 없는 경우, `type`이나 `props`가 변경되었을 때 실제 DOM에 어떻게 반영되나용?

iborymagic replied:
흠. 사실 이 부분은 되도록이면 비교하는 로직에 들어가는 횟수를 최대한 줄이려고 맨 앞에다 배치해놓은 것인데,
콜린의 말대로 children이 없는데 type이나 props가 변경되는 경우를 간과했네요. ㅎㅎ;; ㅋㅋ;;
약간의 위치 조정이 필요할 것 같습니다!!

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  // type 비교
+  if (prevType !== newType) {
+    replaceElement(newVDOM, currentRealDOM);
+    return;
```
vDOM 끼리 비교 후 다른 점이 있다면 바로 RealDOM에 반영하는 구조군요 👍

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  // text node 비교 or children을 타고 재귀 호출
+  newChildren.forEach((newChild, idx) => {
+    if (newChild.type === 'text') {
+      const [prevText] = prevChildren[idx].children;
+      const [newText] = newChild.children;
+
+      if (prevText !== newText) {
+        replaceElement(newVDOM, currentRealDOM);
+        return;
+      }
```
`Array.prototype.forEach`를 사용하면, `100line`에서 리턴을 하더라도 계속해서 루프를 돌텐데 이것도 의도하신걸까용?

iborymagic replied:
엇.. 메모리 이슈..
깜빡하고 forEach문에 return을 사용했네요.
꼼꼼하게 짚어주셔서 감사합니다!!

2SOOY commented on lib/PeactDOM.js:
```diff
@@ -0,0 +1,114 @@
+        replaceElement(newVDOM, currentRealDOM);
+        return;
+      }
+    } else {
+      findDifference(prevChildren[idx], newChild, currentRealDOM.children[idx]);
+    }
```
혹시나 `newChildren`이랑 `prevChildren`의 개수가 다를 때도 정상적으로 작동할까용?
`newChildren`의 개수가 더 많을 경우 마지막 index의 경우 `prevChildren[i]`의 값은 `undefined`가 나올 것 같은 게 추측입니당 ,,ㅎㅎ 
해당 경우에 대한 방어 코드가 있다면 더 안전한 코드가 될 것 같아용 :)

iborymagic replied:
우왕 영양가 넘치는 피드백 정말 감사드립니다.
꼼꼼히 확인하지 못하고 넘어간 부분들을 짚어주셔서 정말 많은 도움이 됐어요.

2SOOY reviewed (commented):
오오오오 !!! `diff`까지 멋지게 구현해주셨군요 👏👏👏
`타입` -> `props` -> `children (with TextContent)` 흐름이 명확하게 보여서 좋았습니다 :)
구현에 대한 설명도 달아주셔서 감사합니다 !!!!

질문 조금 남겼는데, 답변 해주시면 감사하겠습니다 🙏 고생 많으셨어요 피뤄 :)

iborymagic committed:
fix: 콜린 피드백 반영
- realDOM의 children[0]을 firstChild로 수정
- findDifference 함수명 updateDifference로 변경
- children이 없는 경우 return해주는 조건문 위치 이동
- forEach문이 아닌 for문에서 return하도록 수정
- newChildren과 prevChildren이 children 갯수가 다른 경우를 위한 가드 추가

2SOOY reviewed (approved):
깔끔하게 전부 반영해주셨네요 👏👏👏 
미션 수행하시느라 정말 고생 많으셨습니다 피뤄!! 💯💯💯 
다음 단계에서 뵐게요 :)
